{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No son iguales\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "Germplasm = open (\"Germplasm.tsv\" , \"r\")\n",
    "LocusGene = open (\"LocusGene.tsv\" , \"r\") #I Open the tsv files\n",
    "\n",
    "Germplasm.seek(0) \n",
    "LocusGene.seek(0)  #I set the poinnter in the beginning of the file\n",
    "\n",
    "Ger = Germplasm.readlines()\n",
    "Loc = LocusGene.readlines()  #I create two lists withe the information of the lines\n",
    "\n",
    "\n",
    "listG = [] #I create two empty lists to fill them later with the AGI Locus codes\n",
    "listL = []\n",
    "\n",
    "\n",
    "for line1 in Ger[1:]: #With this command, I iterate the lines of the list\n",
    "    Match1 = re.search(r'AT\\dG\\d{5}', line1) #In each line, I search for the AGI Locus Code\n",
    "    listG.append(Match1) #I write the matches (Locus) in the empty list\n",
    "for line2 in Loc[1:]:\n",
    "    Match2 = re.search(r'AT\\dG\\d{5}', line2)\n",
    "    listL.append(Match2)\n",
    "    \n",
    "if listG == listL:        #I compare the two lists to see if they are equal\n",
    "    print(\"Son iguales\")\n",
    "\n",
    "else:\n",
    "    print(\"No son iguales\")\n",
    "\n",
    "Germplasm.close()\n",
    "LocusGene.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this piece of code, I extract the information of the AGI Locus Codes of each file into lists and then I compare them to see if they are the same lists so I can see if the Locus Codes are the same in both files. The answer would be that they are the same so the AGI Locus Code can be used as primary key to relate both tables, but it seems that there is an error in the code so it doesn't print \"Son Iguales\" but I can't see it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database Exam2  #I create and use a database\n",
    "%sql use Exam2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE LocusGene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, LocusLG VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, Protein_Length INTEGER NOT NULL);\n",
    "%sql CREATE TABLE Germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, LocusG VARCHAR(10) NOT NULL, Germplasm VARCHAR(30) NOT NULL, Phenotype VARCHAR(350) NOT NULL, Pubmed INTEGER NOT NULL);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this code, I create in MySQL a database which contains two tables, one for each file. The primary key that relates the two tables will be the AGI Locus Code which is the same in both tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors  #I import the necessary libraries to use pymsql\n",
    "import csv\n",
    "\n",
    "connection = pymysql.connect(host='localhost',       #I connect to the database\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        with open('LocusGene.tsv') as tsvfile:   #I open the tsv file\n",
    "            LocusG = csv.DictReader(tsvfile, delimiter=\"\\t\", quotechar='\"') #In the LocusG object I write the tsv file as a dictionary\n",
    "            for row in LocusG:    #I iterate the dictionary in the object row, then, with the sql command I insert the values of the dictionary in the rows of the table.\n",
    "                sql = \"INSERT INTO LocusGene (LocusLG, Gene, Protein_Length) VALUES ('{}','{}','{}')\".format(row[\"Locus\"],row[\"Gene\"],row[\"ProteinLength\"])\n",
    "                cursor.execute(sql)\n",
    "        with open('Germplasm.tsv') as tsvfile: #I do the same with the second table\n",
    "            Germp = csv.DictReader(tsvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in Germp:\n",
    "                sql = \"INSERT INTO Germplasm (LocusG, Germplasm, Phenotype, Pubmed) VALUES ('{}','{}','{}','{}')\".format(row[\"Locus\"],row[\"germplasm\"],row[\"phenotype\"],row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "        connection.commit()\n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "#print(LocusGene)\n",
    "#print(Germplasm)\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this piece of code, I fill the tables with the information of the tsv files. I use the tsv files as dictionaries and I insert the values of each dictionary in the rows of the tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<function TextIOWrapper.close()>"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM LocusGene INNER JOIN Germplasm ON Germplasm.LocusG = LocusGene.LocusLG;\" \n",
    "        cursor.execute(sql)\n",
    "        Report = cursor.fetchall()\n",
    "finally:\n",
    "    print(\"\")\n",
    "\n",
    "#With the command SELECT, I query the database to obtain the two joined tables. I use inner join to combine tables where the AGI Locus Code is the same \n",
    "#With \"cursor.fecthall()\",I import the results to the Report objec.   \n",
    "try:\n",
    "\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM LocusGene INNER JOIN Germplasm ON Germplasm.LocusG = LocusGene.LocusLG WHERE Gene IN ('SKOR','MAA3');\"\n",
    "        cursor.execute(sql)\n",
    "        Report = Report + cursor.fetchall()\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "#In this query I use the same \"SELECT\" command but I use WHERE to select only the genes SKOR and MAA3\n",
    "\n",
    "\n",
    "with open('finalreport.csv', 'w', newline='') as csvfile:                #I create a csv file and write it with the reports.\n",
    "    reportwriter = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    reportwriter.writerow(Report) \n",
    "csvfile.close"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this code, I import the queries into a csv file. Unfortunately I didn't know how to do the reports 3 and four and only I Know I have to use the commands SELECT COUNT and SELECT AVG, but I don't know how to use them."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
